node/src/components/chain_synchronizer/operations.rs

```diff
@@ -570,54 +559,73 @@ async fn fetch_and_store_block_by_hash(
-    let mut workers: FuturesUnordered<_> = (0..ctx.config.max_parallel_trie_fetches())
-        .map(|worker_id| sync_trie_store_worker(worker_id, abort.clone(), queue.clone(), ctx))
-        .collect();
-    while let Some(result) = workers.next().await {
-        result?; // Return the error if a download failed.
-    }
+    recursive_trie_download(state_root_hash, ctx.clone()).await?;
 
     Ok(())
 }
 
+/// Downloads a given trie and all of its descendants, then stores them into the trie store.
+///
+/// Ensures that all descendant nodes are downloaded using post-order tree traversal, i.e. child
+/// nodes are recursively downloaded and store first before the parent node is stored.
+///
+/// This function does not check for missing descendants for existing tries, the invariant of there
+/// being no trie nodes with missing descendants is assumed.
+#[async_recursion]
+async fn recursive_trie_download(hash: Digest, ctx: Arc<ChainSyncContext>) -> Result<(), Error> {
+    let fetched_trie = fetch_trie_retry_forever(hash, &ctx).await?;
+
+    match fetched_trie {
+        TrieAlreadyPresentOrDownloaded::AlreadyPresent => {
+            // The trie node was already present, meaning all its descendants are also present.
+            // Nothing to do.
+            Ok(())
+        }
+        TrieAlreadyPresentOrDownloaded::Downloaded(trie_bytes) => {
+            let mut descendent_handles: FuturesUnordered<_> =
+                NetTrie::descendants_from_bytes(&trie_bytes)
+                    .map_err(|err| Error::CorruptTrie {
+                        hash,
+                        error: err.to_string(),
+                    })?
+                    .map(|digest| {
+                        // We spawn a new task here instead of awaiting directly to avoid stack
+                        // overflows with deep tries. Furthermore, this automatically parallelizes
+                        // the trie downloads.
+                        tokio::spawn(recursive_trie_download(digest, ctx.clone()))
+                    })
+                    .collect();
+
+            while let Some(result) = descendent_handles.next().await {
+                // Exit early if any of the descendant downloads failed.
+                result.map_err(Error::Join)??;
+            }
+
+            // At this point, we know all descendants are stored, so we store the parent node.
+            let descendants = ctx
+                .effect_builder
+                .put_trie_and_find_missing_descendant_trie_keys(trie_bytes)
+                .await?;
+
+            // Warn if we got an unexpected missing descendent.
+            if !descendants.is_empty() {
+                error!(trie_hash=%hash, "stored trie had missing descendants, even though they were all downloaded. this is a bug or the trie store is corrupted");
+
+                // Note: As an alternative, we can try to recover and download the missing items.
+            }
+
+            Ok(())
+        }
+    }
+}
```

Comment:
It seems like the order in which this would attempt to download the trie would be mostly breadth-first, and we'd end up having most of the trie in memory before storing anything, because the Tokio semaphore will return permits in the order in which they are requested.

We should somehow prioritize deeper trie nodes.